### PR TITLE
[pigeon] added tool to help compare the output from different versions

### PR DESCRIPTION
**Pull request by gaaclarke**
Now that we are not going to check in checked in generated code, this code will help review patches.

## Pre-launch Checklist

- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt.
- [ ] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy].
- [ ] I updated `CHANGELOG.md` to add a description of the change.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the `#hackers-new` channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning

### CODE REVIEW

**Review by stuartmorgan — COMMENTED** [review]
Is there a reason this is bash instead of Dart? Shell scripts are harder to review, harder to test, and don't work on the most popular development platform used by Flutter developers.

**Comment by gaaclarke** [comment]
> Is there a reason this is bash instead of Dart? Shell scripts are harder to review, harder to test, and don't work on the most popular development platform used by Flutter developers.

Yep, it was faster to develop and a fraction of the code.  There is zero branching and it's just a series of command line invocations.  It never would have worked on windows anyways since `diff` doesn't exist there.  Also, it's not critical for working on Pigeon.  It's just a helper script.

File: packages/pigeon/diff_tool.sh [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+xHash=$1
+pigeonPath=$2
+diffTool="diff -ru"
+gitTool="git -c advice.detachedHead=false"
+
+generate_everything() {
+  local inputPath=$1
+  local outputDir=$2
+  pub run pigeon \
+    --input $inputPath \
```
Comment (stuartmorgan) [inline packages/pigeon/diff_tool.sh:25]: `inputPath` needs to be quoted.

File: packages/pigeon/diff_tool.sh [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    --dart_out $outputDir/dart.dart \
```
Comment (stuartmorgan) [inline packages/pigeon/diff_tool.sh:26]: `outputDir` needs to be quoted.

File: packages/pigeon/diff_tool.sh [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+    --java_out $outputDir/java.dart \
+    --objc_header_out $outputDir/objc.h \
+    --objc_source_out $outputDir/objc.m
+}
+
+yHash=$(git rev-parse HEAD)
+xDir=$(mktemp -d -t $xHash)
+yDir=$(mktemp -d -t $yHash)
+inputPath=$yDir/input.dart
+cp $pigeonPath $inputPath
```
Comment (stuartmorgan) [inline packages/pigeon/diff_tool.sh:36]: Both of these need to be quoted.

File: packages/pigeon/diff_tool.sh [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+    --dart_out $outputDir/dart.dart \
+    --java_out $outputDir/java.dart \
+    --objc_header_out $outputDir/objc.h \
+    --objc_source_out $outputDir/objc.m
+}
+
+yHash=$(git rev-parse HEAD)
+xDir=$(mktemp -d -t $xHash)
+yDir=$(mktemp -d -t $yHash)
+inputPath=$yDir/input.dart
```
Comment (stuartmorgan) [inline packages/pigeon/diff_tool.sh:35]: Does `mktemp` guarantee that it won't generate a path containing any spaces? I didn't see any specific guarantee in the docs, but maybe I missed it. If not, these all need to be quoted.

File: packages/pigeon/diff_tool.sh [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+cp $pigeonPath $inputPath
+$gitTool checkout $xHash 1> /dev/null
+generate_everything $inputPath $xDir
+$gitTool $yHash 1> /dev/null
+generate_everything $inputPath $yDir
+$diffTool $yDir $xDir
+rm -rf $yDir
+rm -rf $xDir
```
Comment (stuartmorgan) [inline packages/pigeon/diff_tool.sh:43]: These in particular need to be quoted, because they are extremely dangerous.

**Review by stuartmorgan — CHANGES_REQUESTED** [review]
> it was faster to develop and a fraction of the code

The same can be said of all code without tests; the initial development time isn't the only factor.

> There is zero branching and it's just a series of command line invocations.

It has a function, and does output capture and output redirection. It's not just a trivial "call these 2-3 shell commands for me" script.

Note also that using bash instead of Dart created at least one bug already (incorrect handling of a `pigeonPath` with spaces). So it's already not so trivial that it can't be harder to get right with bash.

> It never would have worked on windows anyways since diff doesn't exist there.

Someone could absolutely install `diff` on Windows. Or this could use `fc` on Windows. It would be trivial to make a Dart version of this script do something useful on Windows, but non-trivial to port the script to PowerShell and maintain two copies.

> Also, it's not critical for working on Pigeon. It's just a helper script.

Apparently it's a useful enough helper that you thought it was worth checking in for people to use.

While I am willing to, reluctantly, approve this as a bash script, I feel strongly that it's the wrong choice. I can think of many, many scripts over the years (including quite a few in Flutter repositories) that were implemented as bash scripts because "it's the easiest option" and "surely we'll never {need to really maintain in/add complexity/want to support Windows}" and where we regretted that decision later. It would not be that hard to write this script in Dart, and the easiest time to write it in Dart is *always* when it's first created, not later when it's potentially become more complex and/or changing it may break people's workflow.

File: packages/pigeon/diff_tool.sh (hunk 1 above)
Comment (gaaclarke) [reply packages/pigeon/diff_tool.sh]: done

File: packages/pigeon/diff_tool.sh (hunk 2 above)
Comment (gaaclarke) [reply packages/pigeon/diff_tool.sh]: done

File: packages/pigeon/diff_tool.sh (hunk 3 above)
Comment (gaaclarke) [reply packages/pigeon/diff_tool.sh]: done

File: packages/pigeon/diff_tool.sh (hunk 4 above)
Comment (gaaclarke) [reply packages/pigeon/diff_tool.sh:35]: Yes, I do believe that is a guarantee as long as you don't do something funky to change $TMPDIR.  Quoted anyways.

File: packages/pigeon/diff_tool.sh (hunk 5 above)
Comment (gaaclarke) [reply packages/pigeon/diff_tool.sh]: Done

**Comment by gaaclarke** [comment]
@stuartmorgan I completely agree and understand.  I think my threshold is a bit higher than yours though.  If I had to throw away 20 lines to rewrite it in Dart I wouldn't protest.  I'll keep an eye for creeping complexity.

**Review by stuartmorgan — APPROVED** [review]
LGTM

> If I had to throw away 20 lines to rewrite it in Dart I wouldn't protest.

Well, I am asking you to do that here, and you are protesting; I'm pretty sure you're not going to be *more* inclined to do it later 🙂

**Pull request merged**